File: sdks/scala/src/main/scala/flapjacksearch/internal/FlapjackAgent.scala
package flapjacksearch.internal

import flapjacksearch.config.AgentSegment

import scala.collection.mutable

/** Handles Flapjack agent segments.
  *
  * An instance of this class maintains a set of [AgentSegment]s, and provides methods to add, remove, and format these
  * segments.
  *
  * @param clientVersion
  *   client version
  */
class FlapjackAgent(clientVersion: String) {
  private val segs = mutable.LinkedHashSet[AgentSegment](
    AgentSegment("Flapjack for Scala", clientVersion),
    AgentSegment("JVM", System.getProperty("java.version"))
  )

  /** Adds a new segment to the agent segments. */
  def addSegment(seg: AgentSegment): FlapjackAgent = {
    if (!segs.contains(seg)) {
      segs += seg
    }
    this
  }

  /** Adds all segments to the agent segments */
  def addSegments(segments: Seq[AgentSegment]): FlapjackAgent = {
    segs.addAll(segments)
    this
  }

  override def toString: String = {
    segs.mkString("; ")
  }
}

object FlapjackAgent {

  /** Creates a new FlapjackAgent instance with the given client version. */
  def apply(clientVersion: String): FlapjackAgent =
    new FlapjackAgent(clientVersion)
}

File: sdks/java/flapjacksearch/src/main/java/com/flapjackhq/config/FlapjackAgent.java
package com.flapjackhq.config;

import java.util.LinkedHashSet;
import java.util.List;
import java.util.Set;
import javax.annotation.Nonnull;

public final class FlapjackAgent {

  private final Set<String> segments;

  private String finalValue;

  public FlapjackAgent(String clientVersion) {
    this.finalValue = String.format("Flapjack for Java (%s)", clientVersion);
    this.segments = new LinkedHashSet<>();
    this.addSegment(new Segment("JVM", System.getProperty("java.version")));
  }

  public FlapjackAgent addSegment(@Nonnull Segment seg) {
    String segment = seg.toString();
    if (!segments.contains(segment)) {
      segments.add(segment);
      finalValue += segment;
    }
    return this;
  }

  public FlapjackAgent addSegments(@Nonnull List<Segment> segments) {
    for (Segment segment : segments) {
      addSegment(segment);
    }
    return this;
  }

  public FlapjackAgent removeSegment(@Nonnull Segment seg) {
    segments.remove(seg.toString());
    return this;
  }

  @Override
  public String toString() {
    return finalValue;
  }

  public static class Segment {

    private final String value;
    private final String version;

    public Segment(String value) {
      this(value, null);
    }

    public Segment(String value, String version) {
      this.value = value;
      this.version = version;
    }

    @Override
    public String toString() {
      StringBuilder sb = new StringBuilder();
      sb.append("; ").append(value);
      if (version != null) {
        sb.append(" (").append(version).append(")");
      }
      return sb.toString();
    }
  }
}
